Clarify QOS class effect (#39118)
kubernetes/website: content/en/docs/concepts/workloads/pods/pod-quality.md

---

> title: Pod Quality of Service Classes

Reviewer: Please consider rename the file to something like `pod-qos.md`. `pod-quality.md` is confusing.

of the {{< glossary_tooltip text="containers" term_id="container" >}} in that Pod, along with
how those requests relate to resource limits.
This is known as {{< glossary_tooltip text="Quality of Service" term_id="qos-class" >}}
(QoS) class. Kubernetes assigns every Pod a QoS class based on the resource requests
and limits of its component {{< glossary_tooltip text="Containers" term_id="container" >}}.
QoS classes are used by Kubernetes to decide which Pods to evict from a Node experiencing
[Node Pressure](/docs/concepts/scheduling-eviction/node-pressure-eviction/). The possible
QoS classes are `Guaranteed`, `Burstable`, and `BestEffort`. When a Node runs out of resources,
Kubernetes will first evict `BestEffort` Pods running on that Node, followed by `Burstable` and
finally `Guaranteed` Pods. When this eviction is due to resource pressure, only Pods exceeding
resource requests are candidates for eviction.

### Guaranteed

Pods that are `Guaranteed` have the strictest resource limits and are least likely
to face eviction. They are guaranteed not to be killed until they exceed their limits
or there are no lower-priority Pods that can be preempted from the Node. They may

> not acquire resources beyond their specified limits. These Nodes can also make

Reviewer: These Nodes --> These Pods

use of exclusive CPUs using the
[`static`](/docs/tasks/administer-cluster/cpu-management-policies/#static-policy) CPU management policy.

#### Criteria

For a Pod to be given a QoS class of `Guaranteed`:

* Every Container in the Pod must have a memory limit and a memory request.
* For every Container in the Pod, the memory limit must equal the memory request.
* Every Container in the Pod must have a CPU limit and a CPU request.
* For every Container in the Pod, the CPU limit must equal the CPU request.

### Burstable

Pods that are `Burstable` have some lower-bound resource guarantees based on the request, but

> do not require a specific limit. If a limit is not specified they default to a

Reviewer: ```suggestion
do not require a specific limit. If a limit is not specified, it defaults to a
```

> limit equivalent to the capacity of the Node, which allows them to flexibly increase

Reviewer: ```suggestion
limit equivalent to the capacity of the Node, which allows the Pods to flexibly increase
```

their resources if resources are available. In the event of Pod eviction due to Node
resource pressure, these Pods are evicted only after all `BestEffort` Pods are evicted.
Because a `Burstable` Pod can include a container that has no resource limits or requests, a Pod
that is `Burstable` can try to use any amount of node resources.

#### Criteria

A Pod is given a QoS class of `Burstable` if:

* The Pod does not meet the criteria for QoS class `Guaranteed`.
* At least one Container in the Pod has a memory or CPU request or limit.

### BestEffort

Pods in the `BestEffort` QoS class can use node resources that aren't specifically assigned
to Pods in other QoS classes. For example, if you have a node with 16 CPU cores available to the
kubelet, and you assign assign 4 CPU cores to a `Guaranteed` Pod, then a Pod in the `BestEffort`
QoS class can try to use any amount of the remaining 12 CPU cores.

The kubelet prefers to evict `BestEffort` Pods if the node comes under resource pressure.

#### Criteria

A Pod has a QoS class of `BestEffort` if it doesn't meet the criteria for either `Guaranteed`

> or `Burstable`. In other words, a Pod is `BestEffort` only if none of the containers in the Pod have a

Reviewer: ```suggestion
or `Burstable`. In other words, a Pod is `BestEffort` only if none of the Containers in the Pod have a
```

> memory limit or a memory request, and none of the containers in the Pod have a

Reviewer: ```suggestion
memory limit or a memory request, and none of the Containers in the Pod have a
```

CPU limit or a CPU request.

> Containers in a Pod can request other resources (not CPU or memory) and still be classed as

Reviewer: ```suggestion
Containers in a Pod can request other resources (not CPU or memory) and still be classified as
```

`BestEffort`.

### Some behavior is independent of QoS class

Certain behavior is independent of the QoS class assigned by Kubernetes. For example:

* Any Container exceeding a resource limit will be killed and restarted by the Kubelet without

> affecting the other Containers in that Pod.

Reviewer: ```suggestion
  affecting other Containers in that Pod.
```

> * If a Container exceeds its resource request and the Node it runs on faces

Reviewer: ```suggestion
* If a Container exceeds its resource request and the node it runs on faces
```

> resource pressure, the Pod it is in becomes a candidate for [eviction](/docs/concepts/scheduling-eviction/node-pressure-eviction/).
> If this occurs all Containers in the Pod will be terminated and rescheduled on a
> different Node.

Reviewer: ```suggestion
  resource pressure, the Pod it is in becomes a candidate for [eviction](/docs/concepts/scheduling-eviction/node-pressure-eviction/).
  If this occurs all Containers in the Pod will be terminated and rescheduled on a
  different Node.
```

* The resource request of a Pod is equal to the sum of the resource requests of

> its component Containers, and the resource limit of a Pod is equal to the sum of
> the resource limits of its component Containers.

Reviewer: ```suggestion
  its component Containers, and the resource limit of a Pod is equal to the sum of
  the resource limits of its component Containers.
```

* The kube-scheduler does not consider QoS class when selecting which Pods to

> [preempt](/docs/concepts/scheduling-eviction/pod-priority-preemption/#preemption).
> Preemption can occur when a cluster does not have enough resources to run all the Pods
> you defined.

Reviewer: ```suggestion
  [preempt](/docs/concepts/scheduling-eviction/pod-priority-preemption/#preemption).
  Preemption can occur when a cluster does not have enough resources to run all the Pods
  you defined.
```

## {{% heading "whatsnext" %}}

> * Learn about [resource management for Pods and Containers](docs/concepts/configuration/manage-resources-containers/).
> 
> * Learn about [Node-pressure eviction](/docs/concepts/scheduling-eviction/node-pressure-eviction/).
> 
> * Learn about [Pod priority and preemption](/docs/concepts/scheduling-eviction/pod-priority-preemption/).
> 
> * Learn about [Pod disruptions](/docs/concepts/workload/pods/disruptions/).
> 
> * Learn how to [assign memory resources to containers and pods](/docs/tasks/configure-pod-container/assign-memory-resource/).
> 
> * Learn how to [assign CPU resources to containers and pods](/docs/tasks/configure-pod-container/assign-cpu-resource/).
> 
> * Learn how to [configure Quality of Service for Pods](/docs/tasks/configure-pod-container/quality-service-pod/).

Reviewer: ```suggestion
* Learn about [resource management for Pods and Containers](docs/concepts/configuration/manage-resources-containers/).
* Learn about [Node-pressure eviction](/docs/concepts/scheduling-eviction/node-pressure-eviction/).
* Learn about [Pod priority and preemption](/docs/concepts/scheduling-eviction/pod-priority-preemption/).
* Learn about [Pod disruptions](/docs/concepts/workload/pods/disruptions/).
* Learn how to [assign memory resources to containers and pods](/docs/tasks/configure-pod-container/assign-memory-resource/).
* Learn how to [assign CPU resources to containers and pods](/docs/tasks/configure-pod-container/assign-cpu-resource/).
* Learn how to [configure Quality of Service for Pods](/docs/tasks/configure-pod-container/quality-service-pod/).
```

---
title: Pod Quality of Service Classes
content_type: concept
weight: 30
---

<!-- overview -->

> This page explains how Kubernetes assigns Quality of Service (QoS) classes to
> Pods as a side effect of specifying resource constraints on Containers. QoS
> classes are by Kubernetes to make decisions about which Pods to evict when
> there are not enough resources on a Node.

Reviewer 2: ```suggestion
This page introduces _Quality of Service (QoS) classes_ in Kubernetes, and explains
how Kubernetes assigns a QoS class to each Pods as a consequence of the resource
constraints that you specify for the containers in that Pod. Kubernetes relies on this
classification to make decisions about which Pods to evict when there are not enough
available resources on a Node.
```

<!-- body -->

## Quality of Service Classes

Because Kubernetes co-locates all Containers within a given {{< glossary_tooltip text="Pod" term_id="pod" >}}
to the same {{< glossary_tooltip text="Node" term_id="node" >}}, it uses a Pod-level
classification to determine how to meet
[resource requirements](docs/concepts/configuration/manage-resources-containers/).
This is known as {{< glossary_tooltip text="Quality of Service" term_id="qos-class" >}}
(QoS) class. Kubernetes assigns every Pod a QoS class based on the resource requests
and limits of its component {{< glossary_tooltip text="Containers" term_id="container" >}}.
QoS classes are used by Kubernetes to decide which Pods to evict from a Node experiencing
[Node Pressure](/docs/concepts/scheduling-eviction/node-pressure-eviction). The possible

> QoS classes are Guaranteed, Burstable, and BestEffort. When a Node runs out of resources,

Reviewer 2: ```suggestion
QoS classes are `Guaranteed`, `Burstable`, and `BestEffort`. When a Node runs out of resources,
```

> Kubernetes will first evict BestEffort Pods running on that Node followed by Burstable and
> Guaranteed Pods. When this eviction is due to resource pressure, only Pods exceeding their

Reviewer 2: ```suggestion
Kubernetes will first evict `BestEffort` Pods running on that Node, followed by `Burstable` and
finally `Guaranteed` Pods. When this eviction is due to resource pressure, only Pods exceeding 
```

resource requests are candidates for eviction.

> 

Reviewer 2: Which is right?
```suggestion
The kube-scheduler also considers QoS class when selecting which Pods to 
[preempt](/docs/concepts/scheduling-eviction/pod-priority-preemption/#preemption) in case
the cluster does not have enough resources to run all the defined Pods.
```
?
```suggestion
However, the kube-scheduler does not consider QoS class when selecting which Pods to 
[preempt](/docs/concepts/scheduling-eviction/pod-priority-preemption/#preemption).
Preemption can occur when a cluster does not have enough resources to run all the Pods
you defined.
```

Reviewer 2: Second one, right?

### Guaranteed

> Guaranteed Pods have the strictest resource limits and the highest priority to avoid

Reviewer 2: ```suggestion
Pods that are `Guaranteed` have the strictest resource limits and are least like to face eviction
```

being evicted. They are guaranteed not to be killed until they exceed their limits
or there are no lower-priotiy Pods that can be evicted from the Node. They may
not acquire resources beyond their specified limits. These Nodes can also make
use of exclusive CPUs using the

> [static](/docs/tasks/administer-cluster/cpu-management-policies/#static-policy) CPU management policy.

Reviewer 2: ```suggestion
[`static`](/docs/tasks/administer-cluster/cpu-management-policies/#static-policy) CPU management policy.
```

#### Criteria

> For a Pod to be given a QoS class of Guaranteed:

Reviewer 2: ```suggestion
For a Pod to be given a QoS class of `Guaranteed`:
```

* Every Container in the Pod must have a memory limit and a memory request.
* For every Container in the Pod, the memory limit must equal the memory request.
* Every Container in the Pod must have a CPU limit and a CPU request.
* For every Container in the Pod, the CPU limit must equal the CPU request.

### Burstable

> Burstable Pods have some lower-bound resource guarantees based on the request, but

Reviewer 2: ```suggestion
Pods that are `Burstable` have some lower-bound resource guarantees based on the request, but
```

do not require a specific limit. If a limit is not specified they default to a
limit equivalent to the capacity of the Node, which allows them to flexibly increase
their resources if resources are available. In the event of Pod eviction due to Node

> resource pressure, these Pods are evicted only after all Best-Effort Pods are evicted.

Reviewer 2: ```suggestion
resource pressure, these Pods are evicted only after all `BestEffort` Pods are evicted.
```

> 

Reviewer 2: ```suggestion
Because a `Burstable` Pod can include a container that has no resource limits or requests, a Pod
that is `Burstable` can try to use any amount of node resources.

```

#### Criteria

> A Pod is given a QoS class of Burstable if:

Reviewer 2: ```suggestion
A Pod is given a QoS class of `Burstable` if:
```

> * The Pod does not meet the criteria for QoS class Guaranteed.

Reviewer 2: ```suggestion
* The Pod does not meet the criteria for QoS class `Guaranteed`.
```

* At least one Container in the Pod has a memory or CPU request or limit.

### BestEffort

> BestEffort Pods have the lowest priority and will be the first Pods evicted if the
> Node runs out of resources, but can use any amount of free resources in the Node.

Reviewer 2: ```suggestion
Pods in the `BestEffort` QoS class can use node resources that aren't specifically assigned
to Pods in other QoS classes. For example, if you have a node with 16 CPU cores available to the
kubelet, and you assign assign 4 CPU cores to a `Guaranteed` Pod, then a Pod in the `BestEffort`
QoS class can try to use any amount of the remaining 12 CPU cores.

The kubelet prefers to evict `BestEffort` Pods if the node comes under resource pressure.
```
```

#### Criteria

> For a Pod to be given a QoS class of BestEffort, the Containers in the Pod must not
> have any memory or CPU limits or requests.

Reviewer 2: ```suggestion
A Pod has a QoS class of `BestEffort` if it doesn't meet the criteria for either `Guaranteed`
or `Burstable`. In other words, a Pod is `BestEffort` only if none of the containers in the Pod have a
memory limit or a memory request, and none of the containers in the Pod have a
CPU limit or a CPU request.
Containers in a Pod can request other resources (not CPU or memory) and still be classed as
`BestEffort`.
```

* The Pod does not meet the criteria for QoS class `Guaranteed`.
* At least one Container in the Pod has a memory or CPU request or limit.

### BestEffort

Pods in the `BestEffort` QoS class can use node resources that aren't specifically assigned
to Pods in other QoS classes. For example, if you have a node with 16 CPU cores available to the
kubelet, and you assign assign 4 CPU cores to a `Guaranteed` Pod, then a Pod in the `BestEffort`
QoS class can try to use any amount of the remaining 12 CPU cores.

The kubelet prefers to evict `BestEffort` Pods if the node comes under resource pressure.

#### Criteria

A Pod has a QoS class of `BestEffort` if it doesn't meet the criteria for either `Guaranteed`
or `Burstable`. In other words, a Pod is `BestEffort` only if none of the Containers in the Pod have a
memory limit or a memory request, and none of the Containers in the Pod have a
CPU limit or a CPU request.
Containers in a Pod can request other resources (not CPU or memory) and still be classified as
`BestEffort`.

### Some behavior is independent of QoS class

Certain behavior is independent of the QoS class assigned by Kubernetes. For example:

> * Any Container exceeding a resource limit will be killed and restarted by the Kubelet without

Reviewer 2: ```suggestion
* Any Container exceeding a resource limit will be killed and restarted by the kubelet without
```

  affecting other Containers in that Pod.

* If a Container exceeds its resource request and the node it runs on faces
  resource pressure, the Pod it is in becomes a candidate for [eviction](/docs/concepts/scheduling-eviction/node-pressure-eviction/).

>   If this occurs all Containers in the Pod will be terminated and rescheduled on a
>   different Node.

Reviewer 2: ```suggestion
  If this occurs, all Containers in the Pod will be terminated. Kubernetes may create a
  replacement Pod, usually on a different node.
```
